src/services/hsm.service.js

```diff
@@ -6,16 +6,208 @@
+   * @returns {Promise<string>}
+   */
+  async getAddressKeyName(index) {
+    if (index in this.addressKeys) {
+      return this.addressKeys[index];
+    }
+
+    const addressKeyName = getBip32Keyname(this.rootKey, 5, { index });
+    const changeKeyName = getChangeKeyname(this.rootKey);
+    try {
+      await this.conn.blockchain.createBip32ChildKeyDerivation(
+        hsm.enums.VERSION_OPTIONS.BIP32_HTR_MAIN_NET,
+        index,
+        false,
+        true,
+        changeKeyName,
+        addressKeyName,
+      );
+    } catch (err) {
+      await delay(500);
+      // (CODE: 5022 | NAME: ERR_OBJ_ALREADY_EXISTS)
+      // This error code means the address key was already derived
+      if (err.errorCode === hsm.constants.ERR_OBJ_ALREADY_EXISTS) {
+        // address key already exists, just return it
+        this.addressKeys[index] = addressKeyName;
+        return addressKeyName;
+      }
+      console.error(err);
+      throw err;
+    }
+
+    this.addressKeys[index] = addressKeyName;
+    return addressKeyName;
+  }
+
+  /**
+   * Sign the given data with the given HSM key
+   * @param {HsmSession} hsmSession
+   * @param {Buffer} dataToSignHash
+   * @param {number} addressIndex
+   * @returns {Promise<Buffer>} DER encoded signature
+   */
+  async signData(dataToSignHash, addressIndex) {
+    // Derive the key to the desired address index
+    const htrKeyName = await this.getAddressKeyName(addressIndex);
+
+    const signature = await this.conn.blockchain.sign(
+      hsm.enums.BLOCKCHAIN_SIG_TYPE.SIG_DER_ECDSA,
+      hsm.enums.BLOCKCHAIN_HASH_MODE.SHA256,
+      dataToSignHash,
+      htrKeyName,
+    );
+
+    // Dinamo SDK returns v + DER, where v is a parity byte
+    // https://manual.dinamonetworks.io/nodejs/enums/hsm.enums.BLOCKCHAIN_SIG_TYPE.html#SIG_DER_ECDSA
+    // We can safely ignore the parity byte
+    if (signature[0] !== 0x30) {
+      return signature.slice(1);
+    }
+    return signature;
+  }
+
+  /**
+   * Sign the given transaction and return the signatures.
+   * @param {hathorLib.Transaction} tx
+   * @param {hathorLib.Storage} storage
+   * @returns {Promise<SignatureData[]>}
+   */
+  async signTx(tx, storage) {
+    const dataToSignHash = tx.getDataToSignHash();
+    /**
+   * @type {SignatureData[]}
+   */
+    const response = [];
+
+    for await (const { tx: spentTx, input, index: inputIndex } of storage.getSpentTxs(tx.inputs)) {
+      const spentOut = spentTx.outputs[input.index];
+
+      if (!spentOut.decoded.address) {
+        // This output is not owned
+        continue;
+      }
+      const addressInfo = await storage.getAddressInfo(spentOut.decoded.address);
+      if (!addressInfo) {
+        // Not a wallet address
+        continue;
+      }
+
+      const pubkeyHex = await storage.getAddressPubkey(addressInfo.bip32AddressIndex);
+      const pubkey = Buffer.from(pubkeyHex, 'hex');
+      const signature = await this.signData(dataToSignHash, addressInfo.bip32AddressIndex);
```

Comment:
```suggestion
      const [pubkeyHex, signature] = await Promise.all([
        storage.getAddressPubkey(addressInfo.bip32AddressIndex),
        this.signData(dataToSignHash, addressInfo.bip32AddressIndex)
      ]);
      const pubkey = Buffer.from(pubkeyHex, 'hex');
```

Reply:
I understand the suggestion but the first promise is just a single derivation from the xpubkey and the other is an api call.
So I don't think we would save a lot of time with this approach.